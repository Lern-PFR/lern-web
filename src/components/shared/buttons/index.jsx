import PropTypes from 'prop-types';
import {
	primary,
	standard,
	subtle,
	dropdown,
	danger,
	iconButton,
	floatingActionButton,
	primaryLink,
	outlined,
	subtleLink,
} from 'theme/buttonStyles';
import DynamicButtonComponent from './DynamicButtonComponent';
import DynamicLinkButtonComponent from './DynamicLinkButtonComponent';

// @TODO: write link and subtleLink buttons once navigation has been set up.

/**
 * @name PrimaryButton
 * @description Button component to use as Call to action. Should only be used once per page (header not included).
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const PrimaryButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...primary} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name PrimaryLinkButton
 * @description Link button component to use as Call to action. Should only be used once per page (header not included).
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const PrimaryLinkButton = ({ children, ...otherProps }) => (
	<DynamicLinkButtonComponent linkStyle={primaryLink} {...primary} {...otherProps}>
		{children}
	</DynamicLinkButtonComponent>
);

/**
 * @name OutlinedButton
 * @description Button component to use as primary action when a CTA is already defined.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const OutlinedButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent linkStyle={primaryLink} {...outlined} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name OutlinedLinkButton
 * @description Link button component to use as primary when a CTA is already defined.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children The children to display inside this button.
 */
const OutlinedLinkButton = ({ children, ...otherProps }) => (
	<DynamicLinkButtonComponent linkStyle={primaryLink} {...outlined} {...otherProps}>
		{children}
	</DynamicLinkButtonComponent>
);

/**
 * @name StandardButton
 * @description Standard button component to be used for most use-cases.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const StandardButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...standard} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name SubtleButton
 * @description Button component to be used along a primary button for less crucial actions such as "cancel".
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const SubtleButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...subtle} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name SubtleLinkButton
 * @description Link button component to be used along a primary button for less used redirections.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const SubtleLinkButton = ({ children, ...otherProps }) => (
	<DynamicLinkButtonComponent linkStyle={subtleLink} {...subtle} {...otherProps}>
		{children}
	</DynamicLinkButtonComponent>
);

/**
 * @name DropdownButton
 * @description Button component to be used for dropdown menu trigger.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const DropdownButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...dropdown} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name DangerButton
 * @description Button component to use for final confirmation to a destructive action such as deleting data.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const DangerButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...danger} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name IconButton
 * @description Button component intended to display only the icon it contains, but behave as a regular button.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const IconButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...iconButton} {...otherProps} isIconOnly>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name FloatingActionButton
 * @description Button component intended to display a floating action button.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const FloatingActionButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...floatingActionButton} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

PrimaryButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

PrimaryLinkButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

OutlinedButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

OutlinedLinkButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

StandardButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

SubtleButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

SubtleLinkButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

DropdownButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

DangerButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

IconButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

FloatingActionButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

export {
	FloatingActionButton,
	OutlinedButton,
	OutlinedLinkButton,
	PrimaryButton,
	PrimaryLinkButton,
	StandardButton,
	SubtleButton,
	SubtleLinkButton,
	DropdownButton,
	DangerButton,
	IconButton,
};
