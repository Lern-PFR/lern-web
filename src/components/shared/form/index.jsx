import FieldsetComponent from './FieldsetComponent';
import LabeledCheckbox from './LabeledCheckbox';
import LabeledRadioButton from './LabeledRadioButton';
import LabelComponent from './LabelComponent';
import LabeledInput from './LabeledInput';

export {
	LabeledCheckbox,
	FieldsetComponent,
	LabeledRadioButton,
	LabelComponent,
	LabeledInput,
};
