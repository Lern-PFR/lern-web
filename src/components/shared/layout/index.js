import MainLayout from './MainLayout';

export {
	// eslint-disable-next-line import/prefer-default-export
	MainLayout,
};
