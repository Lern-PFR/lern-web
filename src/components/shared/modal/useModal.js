import { useCallback, useState, useRef } from 'react';

export default () => {
	const [isShowing, setIsShowing] = useState(false);

	const showingRef = useRef();
	showingRef.current = isShowing;

	const toggle = useCallback(() => setIsShowing(!showingRef.current), [showingRef, setIsShowing]);

	return {
		isShowing,
		toggle,
	};
};
