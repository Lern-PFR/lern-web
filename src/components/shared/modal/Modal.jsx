import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { DoublePica } from 'components/shared/typography';
import { modal, modalBackdrop, modalHeader, modalWrapper } from 'theme/modalStyles';
import { StyledDiv } from '../styledElements';

/**
 * @name Modal
 * @description A generic modal usable inside any page
 *
 * @author Yann Hodiesne
 *
 * @param {node}	children	The components to render inside the modal
 * @param {bool}	isShowing	Determines if the modal is currently being shown to the user
 * @param {string}	title		The title of the modal, displayed inside its header
 */
const Modal = ({ children, isShowing, title }) => {
	if (!isShowing) {
		return null;
	}

	return ReactDOM.createPortal(
		<StyledDiv {...modalBackdrop}>
			<StyledDiv className="modal-wrapper" {...modalWrapper}>
				<StyledDiv className="modal" {...modal}>
					<StyledDiv className="modal-header" {...modalHeader}>
						<DoublePica tag="h2">{title}</DoublePica>
					</StyledDiv>
					<div>
						{children}
					</div>
				</StyledDiv>
			</StyledDiv>
		</StyledDiv>,
		document.body
	);
};

Modal.propTypes = {
	children: PropTypes.node.isRequired,
	isShowing: PropTypes.bool.isRequired,
	title: PropTypes.string.isRequired,
};

export default Modal;
