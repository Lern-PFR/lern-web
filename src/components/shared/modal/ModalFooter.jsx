import PropTypes from 'prop-types';
import { modalFooter } from 'theme/modalStyles';
import { StyledDiv } from '../layout';

/**
 * @name ModalFooter
 * @description Styled div using custom theming properties to be used as footer by Modals.
 *
 * @author Yann Hodiesne
 * @author TimothÃ©e Simon-Franza
 *
 * @param {*} children : The components to render inside the footer.
 */
const ModalFooter = ({ children, ...otherProps }) => (
	<StyledDiv {...modalFooter} {...otherProps}>
		{children}
	</StyledDiv>
);

ModalFooter.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

export default ModalFooter;
