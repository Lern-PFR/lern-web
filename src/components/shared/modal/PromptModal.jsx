import { withTranslation } from 'react-i18next';
import PropTypes from 'prop-types';
import { DangerButton, SubtleButton } from 'components/shared/buttons';
import { BodyCopy } from 'components/shared/typography';
import Modal from './Modal';
import ModalFooter from './ModalFooter';

/**
 * @name PromptModal
 * @description Asks the user for confirmation with the given message
 *
 * @author Yann Hodiesne
 *
 * @param {bool}	isShowing	: Determines if the modal is currently being shown to the user
 * @param {string}	message		: The message displayed to the user
 * @param {func}	confirm		: The function to execute when the user clicks on confirm
 * @param {func}	cancel		: The function to execute when the user clicks on cancel
 * @param {string}	confirmText	: The text displayed inside the confirmation button
 * @param {string}	cancelText	: The text displayed inside the cancel button
 * @param {string}	titleText	: The text displayed inside the confirmation button
 * @param {func}	t			: A translation method provided by the withTranslation HoC
 */
const PromptModal = ({ isShowing, message, confirm, cancel, confirmText, cancelText, titleText, t }) => (
	<Modal isShowing={isShowing} title={titleText || t('components.prompt.title')}>
		<BodyCopy ml="1em">{message}</BodyCopy>
		<ModalFooter>
			<SubtleButton onClick={cancel}>
				{cancelText || t('components.prompt.cancel')}
			</SubtleButton>
			<DangerButton onClick={confirm}>
				{confirmText || t('components.prompt.confirm')}
			</DangerButton>
		</ModalFooter>
	</Modal>
);

PromptModal.propTypes = {
	isShowing: PropTypes.bool.isRequired,
	message: PropTypes.string.isRequired,
	confirm: PropTypes.func.isRequired,
	cancel: PropTypes.func.isRequired,
	confirmText: PropTypes.string,
	cancelText: PropTypes.string,
	titleText: PropTypes.string,
	t: PropTypes.func.isRequired,
};

PromptModal.defaultProps = {
	confirmText: undefined,
	cancelText: undefined,
	titleText: undefined,
};

export default withTranslation()(PromptModal);
