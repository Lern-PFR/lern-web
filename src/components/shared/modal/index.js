import Modal from './Modal';
import ModalFooter from './ModalFooter';
import PromptModal from './PromptModal';
import useModal from './useModal';

export {
	Modal,
	ModalFooter,
	PromptModal,
	useModal,
};
