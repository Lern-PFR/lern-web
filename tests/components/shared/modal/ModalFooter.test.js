import { shallow } from 'enzyme';
import { ModalFooter } from 'components/shared/modal';

describe('ModalFooter', () => {
	it('should render correctly', () => {
		const sut = (
			<ModalFooter>
				<p>Content</p>
			</ModalFooter>
		);

		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});
});
