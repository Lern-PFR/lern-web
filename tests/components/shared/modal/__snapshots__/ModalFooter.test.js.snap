// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalFooter should render correctly 1`] = `
<styled.div
  display="flex"
  justifyContent="flex-end"
  marginTop="1em"
>
  <p>
    Content
  </p>
</styled.div>
`;
